FirebaseMLModelDownloader/Sources/ModelDownloader.swift

```diff
@@ -98,41 +111,203 @@ public class ModelDownloader {
+          modelName: modelName,
+          progressHandler: progressHandler,
+          completion: completion
+        )
+      }
+    case .localModelUpdateInBackground:
+      if let localModel = getLocalModel(modelName: modelName) {
+        DispatchQueue.main.async {
+          completion(.success(localModel))
+        }
+        DispatchQueue.global(qos: .utility).async { [weak self] in
+          self?.getRemoteModel(
+            modelName: modelName,
+            progressHandler: nil,
+            completion: { result in
+              switch result {
+              // TODO: Log outcome of background download
+              case .success: break
+              case .failure: break
+              }
+            }
+          )
+        }
+      } else {
+        getRemoteModel(
+          modelName: modelName,
+          progressHandler: progressHandler,
+          completion: completion
+        )
+      }
 
-    let customModel = CustomModel(
-      name: modelName,
-      size: modelSize,
-      path: modelPath,
-      hash: modelHash
-    )
-    completion(.success(customModel))
-    completion(.failure(.notFound))
+    case .latestModel:
+      getRemoteModel(
+        modelName: modelName,
+        progressHandler: progressHandler,
+        completion: completion
+      )
+    }
   }
 
   /// Gets all downloaded models.
   public func listDownloadedModels(completion: @escaping (Result<Set<CustomModel>,
     DownloadedModelError>) -> Void) {
-    let customModels = Set<CustomModel>()
-    // TODO: List downloaded models
-    completion(.success(customModels))
-    completion(.failure(.notFound))
+    do {
+      let modelPaths = try ModelFileManager.contentsOfModelsDirectory()
+      var customModels = Set<CustomModel>()
+      for path in modelPaths {
+        guard let modelName = ModelFileManager.getModelNameFromFilePath(path) else {
+          completion(.failure(.internalError(description: "Invalid model file name.")))
+          return
+        }
+        guard let modelInfo = getLocalModelInfo(modelName: modelName) else {
+          completion(
+            .failure(.internalError(description: "Failed to get model info for model file."))
+          )
+          return
+        }
+        guard modelInfo.path == path.absoluteString else {
+          completion(
+            .failure(.internalError(description: "Outdated model paths in local storage."))
+          )
+          return
+        }
+        let model = CustomModel(localModelInfo: modelInfo)
+        customModels.insert(model)
+      }
+      completion(.success(customModels))
+    } catch let error as DownloadedModelError {
+      completion(.failure(error))
+    } catch {
+      completion(.failure(.internalError(description: error.localizedDescription)))
+    }
   }
 
   /// Deletes a custom model from device.
   public func deleteDownloadedModel(name modelName: String,
                                     completion: @escaping (Result<Void, DownloadedModelError>)
```

Comment:
Hmm, I may have missed this in the public API review but it seems like this shouldn't be a result, but should be an optional DownloadModelError as the return type. WDYT?

Reply:
Yes we talked about this and went with Result to keep it consistent with all the other public APIs. We could change it to an optional if needed, though.

Reply:
Something to revisit later I think, not a blocker for now obviously